Fixed #24816 -- Clarified docs about preventing duplicate signals. (#13131)
django/django: docs/topics/signals.txt

multiple times. This can cause your receiver function to be registered more
than once, and thus called multiple times for a single signal event.

As Django has evolved, this issue has generally less risk of presenting itself.

> Before ``AppConfig.ready`` was introduced to Django, it was possible that

Reviewer: Add which version this change was made? I think it helps add some more certainty to the time frame in which Django has 'evolved'.

Author: I'm not sure we should necessarily say the issue is "gone", since `AppConfig.ready` can be called multiple times during django's setup flow. I was mostly following through the instruction by the ticket (e.g. specifying that ready can be called multiple times, but at least that's a "standard" entrypoint that you can plan for). Maybe we should remove/reword this? Assuming the counter point is valid.

Author: Resolving for now as the wording has changed (and I've removed this idea/reference).

during the Django setup process during normal operation or from a management

> command, the handlers.py file would be imported multiple times, and thus,

Reviewer: - Suggestion to add a full stop after 'multiple time'. (I think this long sentence needs breaking up to help readability, here is where I think a break could take place with the current wording)

- Suggestion to change 'thus' to 'so' or 'therefore'

Author: Tried out something like that on my change. Please let me know!

the signal handler would be registered each time.

> Another (less common but still possible) occurrence is if in your application
> code you import your handlers.py module from somewhere in your project because

Reviewer: ```suggestion
Another potential occurrence is, if your application imports
handlers.py from somewhere in your project because
```

Author: I've made an attempt to reduce the wording and clarity in general. Resolving for now as the words are probably out of date now, sorry!

> you need to call or run logic against a handler function, and thus the signal
> registration would run at least one more time per-import instance. As projects

Reviewer: ```suggestion
you need to call or run logic against a handler function. This will result in the signal
registration being run at least one more time per-import instance. As projects
```

Author: Hopefully the new wording is clearer, but resolving since the sentence has changed.

tend to grow larger over time (more features, bug fixes), this issue could
present itself, regardless of ``AppConfig.ready`` being part of Django.

> 

Reviewer: Remove blank line

Author: right  on!

multiple times. This can cause your receiver function to be registered more
than once, and thus called multiple times for a single signal event.

> Although :meth:`~django.apps.AppConfig.ready` is designed to be called

Reviewer 2: I think we should move this note after: _"In some circumstances, the code connecting receivers to signals may runmultiple times. "_, e.g.
```diff
diff --git a/docs/topics/signals.txt b/docs/topics/signals.txt
index 30c8299fae..261027932c 100644
--- a/docs/topics/signals.txt
+++ b/docs/topics/signals.txt
@@ -187,8 +187,12 @@ Preventing duplicate signals
 ----------------------------
 
 In some circumstances, the code connecting receivers to signals may run
-multiple times. This can cause your receiver function to be registered more
-than once, and thus called multiple times for a single signal event.
+multiple times. For example, the :meth:`~django.apps.AppConfig.ready` method
+may be executed more than once during testing, another cause is if you import
+the module where you define signals from somewhere in your project (the signal
+registration will run as many times as it is imported). This can cause your
+receiver function to be registered more than once, and thus called multiple
+times for a single signal event.
 
 If this behavior is problematic (such as when using signals to
 send an email whenever a model is saved), pass a unique identifier as
```

Author: I gave it a try. I'm finding it's getting challenging to decide exactly how to express and describe this (gone through three or four drafts), but I agree in the majority of what you've recommended. Resolving for now, since I've changed the wording. Please review the newest iteration (there is some duplication of wording/concept, but I'm unsure that it's a bad thing).

only once during Django's startup sequence, unit test runs may break this
implicit contract. Another (albeit less likely) cause for this is if

> you import your handlers.py module from somewhere in your project; the signal

Reviewer 3: `handlers.py` is not _necessarily_ where such things exist. A vague convention might exist (but could also just as easily be `listeners.py` or `signals.py` if someone so wished) but it's not a requirement, and whether or not those imports are OK from the point of view of signal registration depends entirely if they perform their _registration_ there, or just _exist_ there (precisely because registration can and probably should be deferred until the AppConfig's `ready()` method is called).

Or, to summarise. Don't say `handlers.py` (as far as I can tell, it's not mentioned elsewhere in this document)

Author: Looks like there is wording related to `signals` submodule (as a convention), but good call -- it isn't worth duplicating/binding/confusing users, when other parts of the docs already make suggestions (not assumptions). I've made the update.

the signal registration will run as many times as it is imported.
As projects tend to grow larger over time (more features, bug fixes),

> this issue could present itself.

Reviewer 2: I would chop this.

Author: Yeah, it's a "software engineering" honest moment, but agreed, it doesn't add anything explicit or necessary.
